src/mame/machine/hp2640_tape.cpp

```diff
@@ -0,0 +1,547 @@
+		// Acceleration: 2000 in/s^2
+		finder->set_acceleration(ACCELERATION);
+		// Slow speed: 10 ips, Fast speed: 60 ips
+		finder->set_set_points(SLOW_SPEED , FAST_SPEED);
+		// 58.4 ticks per inch
+		finder->set_tick_size(TACH_TICK_LENGTH);
+		// Manchester encoded data
+		finder->set_image_format(hti_format_t::HTI_MANCHESTER_MOD);
+		// Moving when speed is >1 ips
+		finder->set_go_threshold(MOVING_THRESHOLD);
+
+		finder->hole().set([this , i](int state) { hole_w(i , state); });
+		finder->tacho_tick().set([this , i](int state) { tacho_tick_w(i , state); });
+		finder->motion_event().set([this , i](int state) { motion_w(i , state); });
+		finder->rd_bit().set([this , i](int state) { rd_bit_w(i , state); });
+		finder->wr_bit().set([this , i](int state) { return wr_bit_r(i); });
+	}
+}
+
+void hp2640_tape_device::device_start()
+{
+	m_irq_handler.resolve_safe();
+	m_led0_handler.resolve_safe();
+	m_led1_handler.resolve_safe();
+
+	m_gap_timer = timer_alloc(GAP_TMR_ID);
+	m_cell_timer = timer_alloc(CELL_TMR_ID);
+}
+
+void hp2640_tape_device::device_reset()
+{
+	m_selected_drive = 0;
+
+	// All bits set to 1 but 3 & 0
+	m_cmd_reg = ~0;
+	BIT_CLR(m_cmd_reg , CMD_REG_WRITE_BIT);
+	BIT_CLR(m_cmd_reg , CMD_REG_RUN_BIT);
+
+	m_modulus = MODULUS_RESET;
+	m_cell_cnt = 6;
+	m_bit_cnt = 8;
+	m_tach_div2 = false;
+	m_tach_latch = false;
+	m_hole_latch = false;
+	m_byte_ready = false;
+	m_irq = true;
+	m_bit_sync = false;
+	m_isf = false;
+	m_gap = true;
+	m_prev_gap = true;
+	m_rip = false;
+	m_current_op = hp_dc100_tape_device::OP_READ;
+	update_irq();
+
+	m_gap_timer->reset();
+	m_cell_timer->reset();
+}
+
+void hp2640_tape_device::device_timer(emu_timer &timer, device_timer_id id, int param, void *ptr)
+{
+	LOG("TMR %d @%s\n" , id , machine().time().to_string());
+
+	switch (id) {
+	case GAP_TMR_ID:
+		m_gap = !m_gap;
+		if (m_gap) {
+			set_gap();
+		} else {
+			LOG("GAP ENDS\n");
+			load_gap_timer();
+			if (m_isf) {
+				load_modulus();
+			}
+		}
+		break;
+
+	case CELL_TMR_ID:
+		if (m_current_op == hp_dc100_tape_device::OP_READ) {
+			m_cell_cnt++;
+			if (m_cell_cnt == 15) {
+				restart_cell_cnt();
+			} else {
+				load_modulus();
+			}
+		}
+		break;
+
+	default:
+		break;
+	}
+}
+
+void hp2640_tape_device::hole_w(unsigned drive , int state)
+{
+	if (state && drive == m_selected_drive) {
+		LOG("HOLE\n");
+		m_hole_latch = true;
+		BIT_CLR(m_cmd_reg, CMD_REG_RUN_BIT);
+		set_speed();
+		update_irq();
+	}
+}
+
+void hp2640_tape_device::tacho_tick_w(unsigned drive , int state)
+{
+	if (state && drive == m_selected_drive) {
+		m_tach_div2 = !m_tach_div2;
+		if (m_tach_div2) {
+			LOG("HALF TICK\n");
+			m_tach_latch = true;
+			update_irq();
+		}
+	}
+}
+
+void hp2640_tape_device::motion_w(unsigned drive , int state)
+{
+	if (state && drive == m_selected_drive) {
+		LOG("MOTION\n");
+		start_rd_wr();
+	}
+}
+
+void hp2640_tape_device::rd_bit_w(unsigned drive , int state)
+{
+	if (drive == m_selected_drive) {
+		bool rd_bit = !!state;
```

Comment:
We tend to use `bool(x)` or `x != 0` to do these, I’d rather we don’t get another convention for converting an integer to Boolean proliferating.